### DISCUSSION TITLE
PDF/A-3

### DISCUSSION THREAD

**Discussion by reiz** [prawnpdf/prawn#1335] [General] [answered]
Does prawn support PDF/A-3? I would like to take a regular PDF document, convert it to PDF/A-3 and add an XML file to it. Is that possible with prawn?

**Comment by pointlessone** [comment, marked as answer]
No, Prawn doesn't support any of the PDF/A profiles. You can use low-level pdf-core APIs to add XML to a document.

**Reply by reiz** [reply]
@pointlessone thx for the fast response. Which low-level pdf-core API would you recommend? Any specific Gem?

**Reply by gettalong** [reply]
Prawn uses the pdf-core gem for the low-level API functionality, see https://github.com/prawnpdf/pdf-core/

**Reply by reiz** [reply]
@pointlessone @gettalong In the soon future, invoices in the EU will have to be PDF/A-3 documents with embedded XML. There are different regulatory standards which come together, e.g. ZUGFeRD, xRechnung, and Facutur-X. Just call the most important ones. If prawn can not support PDF/A-3 it will be pretty much useless for invoice creation. 
But even without this regulatory nightmare, many big companies require invoices to be at least PDF/A-1.
It would super awesome if prawn would somehow support PDF/A-1 & PDF/A-3.

**Reply by gettalong** [reply]
@reiz Thanks for the pointer! I had a quick look at XRechnung and as far as I can see there is no requirement as to the format of PDF. Indeed, the XRechnung specification at https://xeinkauf.de/app/uploads/2023/02/231-XRechnung-2023-02-03.pdf says in section 6.1 that the XRechnung *must not* be embedded into a PDF because if it is it doesn't conform to the standard.

The only thing I found is that if you want to comply with ZUGFeRD, then you would need PDF/A-3 and the XML file within.

I agree that PDF/A support would generally be a good thing. Note, however, that taking an existing file and making it a valid PDF/A file is *not* easily possible, even with a full-blown PDF library (e.g. if you need to embed some non-embedded fonts). So, if at all, this would probably only be supported for PDF documents created with Prawn.

**Reply by reiz** [reply]
@gettalong You are right. It's mainly ZUGFeRD and Factur-X. If Prawn would support PDF/A-3 with embedded XML files, for PDFs which are created by Prawn, it would be wonderful!! In that case, I would switch from PDFKit to Prawn. :-)

**Reply by pointlessone** [reply]
We'd be delighted to get PDF/A support contributed.

**Reply by reiz** [reply]
@pointlessone @gettalong I looked into the [PDFA specs](https://pdfa.org/resource/iso-19005-pdfa/) and especially into the [PDF/A-1 specs](https://pdfa.org/resource/technical-note-tn0003-metadata-in-pdfa-1/). The last one is not that complicated and should be relatively easy ... for somebody who has already experience with PDF creation/manipulation ;-) 
I did already some experiments for adding XMP metadata. But not successful yet.
